daceml/onnx/op_implementations/pure_implementations.py

```diff
@@ -0,0 +1,617 @@
+                           dace.Memlet.simple(tmp_out, 'i, j'))
+        state_exp.add_edge(mx1, None, tmp_out, None,
+                           dace.Memlet.simple(tmp_out, '0:' + I + ', 0:' + J))
+
+        me2, mx2 = state_exp.add_map('map2', dict(i='0:' + I, j='0:' + J))
+        state_exp.add_edge(tmp_out, None, me2, None,
+                           dace.Memlet.simple(tmp_out, '0:' + I + ', 0:' + J))
+        state_exp.add_edge(me2, None, task2, '_a2',
+                           dace.Memlet.simple(tmp_out, 'i, j'))
+        state_exp.add_edge(task2, '_b2', mx2, None,
+                           dace.Memlet.simple(output, 'i, j'))
+        state_exp.add_edge(mx2, None, output, None,
+                           dace.Memlet.simple(output, '0:' + I + ', 0:' + J))
+        sdfg_exp.fill_scope_connectors()
+
+        return sdfg_exp
+
+
+@autoregister_params(op="ReduceSum", name="pure")
+class PureReduceSum(ONNXForward):
+    @staticmethod
+    def forward(node: ONNXOp, state: SDFGState,
+                sdfg: SDFG) -> typing.Union[Node, SDFG]:
+        node.validate(sdfg, state)
+
+        axes = node.axes
+
+        # when keepdims is true, this works but there is a useless copy. We just leave this for now; this can be fixed
+        # with a reshape node when those exist.
+        def prog(data, reduced):
+            reduced[:] = np.sum(data, axis=axes)
+
+        return program_for_node(prog, sdfg, state, node).to_sdfg()
+
+
+@autoregister_params(op="ReduceMax", name="pure")
+class PureReduceMax(ONNXForward):
+    @staticmethod
+    def forward(node: ONNXOp, state: SDFGState,
+                sdfg: SDFG) -> typing.Union[Node, SDFG]:
+        node.validate(sdfg, state)
+
+        axes = node.axes
+
+        # when keepdims is true, this works but there is a useless copy. We just leave this for now; this can be fixed
+        # with a reshape node when those exist.
+        def prog(data, reduced):
+            reduced[:] = np.max(data, axis=axes)
+
+        return program_for_node(prog, sdfg, state, node).to_sdfg()
+
+
+@autoregister_params(op="ReduceMin", name="pure")
+class PureReduceMin(ONNXForward):
+    @staticmethod
+    def forward(node: ONNXOp, state: SDFGState,
+                sdfg: SDFG) -> typing.Union[Node, SDFG]:
+        node.validate(sdfg, state)
+
+        axes = node.axes
+
+        # when keepdims is true, this works but there is a useless copy. We just leave this for now; this can be fixed
+        # with a reshape node when those exist.
+        def prog(data, reduced):
+            reduced[:] = np.min(data, axis=axes)
+
+        return program_for_node(prog, sdfg, state, node).to_sdfg()
+
+
+@autoregister_params(op="Softmax", name="pure")
+class PureSoftmax(ONNXForward):
+    @staticmethod
+    def forward(node: ONNXOp, state: SDFGState,
+                sdfg: SDFG) -> typing.Union[Node, SDFG]:
+
+        # NOTE: once there is a reshape node this whole expansion becomes much simpler:
```

Comment:
I don't think a reshape node is entirely necessary (though it's coming anyway), but some implementation of keepdims. In any case, we can keep it this way until the reshape.